pkg/ebpf/c/tracee.bpf.c

```diff
@@ -4861,6 +4861,104 @@ int BPF_KPROBE(trace_ret_exec_binprm2)
     return events_perf_submit(&p, PROCESS_EXECUTION_FAILED, ret);
 }
 
+// Submit "access_remtoe_vm" event for the given vma and info.
+statfunc int submit_access_remote_vm(program_data_t *p,
+                                     struct vm_area_struct *vma,
+                                     void *address,
+                                     unsigned int gup_flags)
+{
+    unsigned long vm_flags = get_vma_flags(vma);
+    struct mm_struct *mm = BPF_CORE_READ(vma, vm_mm);
+    file_info_t file_info = {};
+
+    char location[64] = {0};
+
+    // Get mapped memory information
+    struct file *file = get_mapped_file_from_vma(vma);
+    file_info = get_file_info(file);
+    // If there is no file that is mapped to the vma, try to get memory location name instead.
+    if (!file_info.pathname_p) {
+        get_vma_location(vma, sizeof(location), location);
+    }
+
+    struct task_struct *task = get_owner_task_from_mm(mm);
+    u32 pid = get_task_host_tgid(task);
+
+    save_to_submit_buf(&p->event->args_buf, &pid, sizeof(u32), 0);
+    save_to_submit_buf(&p->event->args_buf, &address, sizeof(void *), 1);
+    save_to_submit_buf(&p->event->args_buf, &gup_flags, sizeof(unsigned int), 2);
+    save_to_submit_buf(&p->event->args_buf, &vm_flags, sizeof(unsigned long), 3);
+    if (file_info.pathname_p != NULL) {
+        save_str_to_buf(&p->event->args_buf, file_info.pathname_p, 4);
+        save_to_submit_buf(&p->event->args_buf, &file_info.id.device, sizeof(dev_t), 5);
+        save_to_submit_buf(&p->event->args_buf, &file_info.id.inode, sizeof(unsigned long), 6);
+        save_to_submit_buf(&p->event->args_buf, &file_info.id.ctime, sizeof(u64), 7);
+    } else {
+        save_str_to_buf(&p->event->args_buf, location, 4);
+    }
+
+    return events_perf_submit(p, ACCESS_REMOTE_VM, 0);
+}
+
+SEC("kprobe/generic_access_phys")
```

Comment:
Okay, this **could** be problematic if we run workloads doing userland MMAP IO to PCI devices, for example (an environment with GPUs through PCI, maybe ? Don't know). I think we should add a note here. If you check, the `vm_operations->access` function pointer can be set to `generic_access_phys` at some points, and, besides the char device driver responsible for /proc/pid/mem, there is also the MMIO code for PCI devices (when mapped in userland to bypass kernel). Again, I don't know if that would be true, but a note here for further awareness would be good.

Reply:
I will write it in a comment, but from what I have seen the `access` method isn't called by many flows in the kernel, so I don't think the load will be significant.
For example, in kernel 5.16 it is used only in  `__access_remote_vm` and in `kernfs_vma_access`, which is also an implementation of the `vma_ops->access`.
Notice that even in the declaration of the struct, it is written that this method is only called by `__access_remote_vm`:
```C
	/* called by access_process_vm when get_user_pages() fails, typically
	 * for use by special VMAs. See also generic_access_phys() for a generic
	 * implementation useful for any iomem mapping.
	 */
	int (*[access](https://elixir.bootlin.com/linux/latest/C/ident/access))(struct [vm_area_struct](https://elixir.bootlin.com/linux/latest/C/ident/vm_area_struct) *[vma](https://elixir.bootlin.com/linux/latest/C/ident/vma), unsigned long addr,
		      void *buf, int len, int [write](https://elixir.bootlin.com/linux/latest/C/ident/write));
```
So I don't think we need to be concerned about the performance here more than in `__access_remote_vm`.